samples/nrf9160/nrf_cloud_mqtt_multi_service/src/connection.c

```diff
@@ -0,0 +1,648 @@
+}
+
+
+bool await_date_time_known(k_timeout_t timeout)
+{
+	return k_event_wait(&cloud_connection_events, DATE_TIME_KNOWN, false, timeout);
+}
+
+/**
+ * @brief Notify that connection to nRF Cloud has been established
+ */
+static void notify_cloud_connected(void)
+{
+	k_event_post(&cloud_connection_events, CLOUD_CONNECTED);
+}
+
+/**
+ * @brief Notify that cloud connection is ready
+ */
+static void notify_cloud_ready(void)
+{
+	k_event_post(&cloud_connection_events, CLOUD_READY);
+}
+
+/**
+ * @brief Clear nRF Cloud connection events
+ */
+static void clear_cloud_connection_events(void)
+{
+	k_event_set(&cloud_connection_events, 0);
+}
+
+/**
+ * @brief Await a connection to nRF Cloud (ignoring LTE connection state and cloud readiness)
+ *
+ * @param timeout - The time to wait before timing out
+ * @return true if occurred
+ * @return false if timed out
+ */
+
+static bool await_cloud_connected(k_timeout_t timeout)
+{
+	LOG_DBG("Awaiting Cloud Connection");
+	return k_event_wait(&cloud_connection_events, CLOUD_CONNECTED, false, timeout);
+}
+
+
+/**
+ * @brief Wait for nRF Cloud readiness (fully connected and ready, ignoring LTE state)
+ *
+ * @param timeout - The time to wait before timing out
+ * @return true if occurred
+ * @return false if timed out
+ */
+static bool await_cloud_ready(k_timeout_t timeout)
+{
+	LOG_DBG("Awaiting Cloud Ready");
+	return k_event_wait_all(&cloud_connection_events, CLOUD_CONNECTED | CLOUD_READY,
+				false, timeout);
+}
+
+bool await_connection(k_timeout_t timeout)
+{
+	return await_lte_connection(timeout) && await_cloud_ready(timeout);
+}
+
+bool cloud_is_connected(void)
+{
+	return k_event_wait(&cloud_connection_events, CLOUD_CONNECTED, false, K_NO_WAIT);
+}
+
+void disconnect_cloud(void)
+{
+	k_event_post(&cloud_connection_events, CLOUD_DISCONNECTED);
+}
+
+bool await_cloud_disconnection(k_timeout_t timeout)
+{
+	return k_event_wait(&cloud_connection_events, CLOUD_DISCONNECTED, false, timeout);
+}
+
+bool cloud_is_disconnecting(void)
+{
+	return k_event_wait(&cloud_connection_events, CLOUD_DISCONNECTED, false, K_NO_WAIT);
+}
+
+/**
+ * @brief Handler for date_time events. Used exclusively to detect when we have obtained
+ * a valid modem time.
+ *
+ * @param evt
+ */
+static void date_time_evt_handler(const struct date_time_evt *evt)
+{
+	if (date_time_is_valid()) {
+		notify_date_time_known();
+	}
+}
+
+/**
+ * @brief Handler for events from nRF Cloud Lib.
+ *
+ * @param nrf_cloud_evt Passed in event.
+ */
+static void cloud_event_handler(const struct nrf_cloud_evt *nrf_cloud_evt)
+{
+	switch (nrf_cloud_evt->type) {
+	case NRF_CLOUD_EVT_TRANSPORT_CONNECTED:
+		LOG_DBG("NRF_CLOUD_EVT_TRANSPORT_CONNECTED");
+		/* Notify that we have connected to the nRF Cloud */
+		notify_cloud_connected();
+		break;
+	case NRF_CLOUD_EVT_TRANSPORT_CONNECTING:
+		LOG_DBG("NRF_CLOUD_EVT_TRANSPORT_CONNECTING");
+		break;
+	case NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST:
+		LOG_DBG("NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST");
```

Comment:
Ah, no, I have not tested this yet. The connection logic _**should**_ handle this, but very slowly, and via a fail-safe. It would be a good idea to handle this scenario directly, thank you for pointing it out!

Reply:
Okay, turns out the connection process was slightly out-of-spec with what the MQTT API expects. I've got that corrected, now, I think about as cleanly as is possible with this simplified connect loop approach.